### ISSUE TITLE
[JENKINS-32266] Priority is not respected for downstream jobs triggered automatically

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/priority-sorter-plugin#589]
<h1><a name="Description"></a>Description</h1>

<p>If jobs are added to Build Queue by starting them manually, they are added to the list in the required order (according to priorities). However, this is not the case when jobs are triggered automatically based on upstream/downstream relationship - they are re-sorted only after some observable timeout:</p>
<ul>
	<li>If new job is added to Build Queue by triggering it from upstream one, the new job stays on top of Build Queue job list (the last to run) even if its priority is <em>higher</em> (priority number is <em>smaller</em>) than any other job in the Build Queue. Refreshing the main page continuously shows the same job list order.</li>
	<li>After some timeout, refreshing the main page finally re-sorts the job list in the Build Queue according to priorities.</li>
</ul>

<p>In order to easily observe this behaviour, use quiet down mode ("Manage Jenkins" => "Prepare for Shutdown") to keep jobs in the Build Queue without running them.</p>

<h1><a name="Stepstoreproduce"></a>Steps to reproduce</h1>

<ul class="alternate" type="square">
	<li>Create 2 priority groups:<br/>
  high_priority_group (e.g. priority number = 3)<br/>
  low_prirority_group (e.g. priority number = 5)</li>
</ul>

<ul class="alternate" type="square">
	<li>Create 3 jobs:<br/>
  low_a (priority = low_priority_group)<br/>
  high_b (priority = high_priority_group)<br/>
  high_c (priority = high_priority_group)</li>
</ul>

<ul class="alternate" type="square">
	<li>Configure job high_c as downstream of high_b. In other words, when high_b competes, high_c is supposed to be started.</li>
</ul>

<ul class="alternate" type="square">
	<li>Set quiet down mode ("Manage Jenkins" => "Prepare for Shutdown").</li>
</ul>

<ul class="alternate" type="square">
	<li>Schedule both low_a and high_b jobs in any order manually. Build Queue will list low_a first and high_b second as expected (high_b to be run first). Refresh Jenkins main page to confirm this.</li>
</ul>

<ul class="alternate" type="square">
	<li>Cancel quiet down mode for a while to get high_b job started. Set quiet down mode again to be able to see Build Queue sorting mode when high_c is triggered automatically.</li>
</ul>

<ul class="alternate" type="square">
	<li>As soon as high_c is triggered automatically, it is first listed at the top of the Build Queue (in a wrong order) which says low_a is to be run first. Refresh Jenkins main page to confirm this.</li>
</ul>

<ul class="alternate" type="square">
	<li>After a while refreshing Jenkins main page re-sorts jobs in the Build Queue to correct order.</li>
</ul>

<p><b><font color="red">The problem is that without quiet down mode, the low_a job with lower priority will start ahead of high_c job (which is triggered automatically by upstream high_b job).</font></b></p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=uvsmtid">uvsmtid</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-32266" target="_blank">Priority is not respected for downstream jobs triggered automatically</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/emsa23.png" title="emsa23's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=emsa23">emsa23</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Critical
<li><b>component(s)</b>: prioritysorter-plugin
<li><b>resolution</b>: Not A Defect
<li><b>resolved</b>: 2016-01-21T14:55:01+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 2
<li><b>imported</b>: 20251211-071809
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<h1><a name="Description"></a>Description</h1>

<p>If jobs are added to Build Queue by starting them manually, they are added to the list in the required order (according to priorities). However, this is not the case when jobs are triggered automatically based on upstream/downstream relationship - they are re-sorted only after some observable timeout:</p>
<ul>
	<li>If new job is added to Build Queue by triggering it from upstream one, the new job stays on top of Build Queue job list (the last to run) even if its priority is <em>higher</em> (priority number is <em>smaller</em>) than any other job in the Build Queue. Refreshing the main page continuously shows the same job list order.</li>
	<li>After some timeout, refreshing the main page finally re-sorts the job list in the Build Queue according to priorities.</li>
</ul>

<p>In order to easily observe this behaviour, use quiet down mode ("Manage Jenkins" =&gt; "Prepare for Shutdown") to keep jobs in the Build Queue without running them.</p>

<h1><a name="Stepstoreproduce"></a>Steps to reproduce</h1>

<ul class="alternate" type="square">
	<li>Create 2 priority groups:
  high_priority_group (e.g. priority number = 3)
  low_prirority_group (e.g. priority number = 5)</li>
</ul>

<ul class="alternate" type="square">
	<li>Create 3 jobs:
  low_a (priority = low_priority_group)
  high_b (priority = high_priority_group)
  high_c (priority = high_priority_group)</li>
</ul>

<ul class="alternate" type="square">
	<li>Configure job high_c as downstream of high_b. In other words, when high_b competes, high_c is supposed to be started.</li>
</ul>

<ul class="alternate" type="square">
	<li>Set quiet down mode ("Manage Jenkins" =&gt; "Prepare for Shutdown").</li>
</ul>

<ul class="alternate" type="square">
	<li>Schedule both low_a and high_b jobs in any order manually. Build Queue will list low_a first and high_b second as expected (high_b to be run first). Refresh Jenkins main page to confirm this.</li>
</ul>

<ul class="alternate" type="square">
	<li>Cancel quiet down mode for a while to get high_b job started. Set quiet down mode again to be able to see Build Queue sorting mode when high_c is triggered automatically.</li>
</ul>

<ul class="alternate" type="square">
	<li>As soon as high_c is triggered automatically, it is first listed at the top of the Build Queue (in a wrong order) which says low_a is to be run first. Refresh Jenkins main page to confirm this.</li>
</ul>

<ul class="alternate" type="square">
	<li>After a while refreshing Jenkins main page re-sorts jobs in the Build Queue to correct order.</li>
</ul>

<p><b><font color="red">The problem is that without quiet down mode, the low_a job with lower priority will start ahead of high_c job (which is triggered automatically by upstream high_b job).</font></b></p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Jenkins ver. 1.641<br/>
Priority Sorter Plugin 3.4
```
</details>
<details><summary><i>1 attachment</i></summary>

- [PrioritySorter.log](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/31647/PrioritySorter.log)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/emsa23.png" title="emsa23's avatar" /> emsa23:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-32266?focusedId=246573&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-246573">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>As I said above it is potentially possible to do something like that but that will result in blocked jobs and unused workers and the point of the plugin is to maximize the usage of the workers for prioritized and important jobs not to enforce strict order.</p>

<p>Adding checks on waiting jobs will not solve all problems for you. As you can see in your logs the job in the queue can for example be started before the the new job even becomes waiting. Also consider the case when you have many workers that will build in parallell even if the jobs are in the 'correct' order they may be started almost at the same time on different workers.</p>

<p>The best way to solve the order issue where one jobs needs to be done before the next starts is not to trigger the next jobs until it is allowed to run and/or to use pipeline/workflow plugins available and block downstream jobs until they are allowed to run.</p>
</pre>
</details>

</ul>
</details>

<p>As I said above it is potentially possible to do something like that but that will result in blocked jobs and unused workers and the point of the plugin is to maximize the usage of the workers for prioritized and important jobs not to enforce strict order.</p>

<p>Adding checks on waiting jobs will not solve all problems for you. As you can see in your logs the job in the queue can for example be started before the the new job even becomes waiting. Also consider the case when you have many workers that will build in parallell even if the jobs are in the 'correct' order they may be started almost at the same time on different workers.</p>

<p>The best way to solve the order issue where one jobs needs to be done before the next starts is not to trigger the next jobs until it is allowed to run and/or to use pipeline/workflow plugins available and block downstream jobs until they are allowed to run.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/emsa23.png" title="emsa23's avatar" /> emsa23:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-32266?focusedId=246574&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-246574">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Closing this issue as this is not a defect the plugin works as expected.</p>

<p>If you like a change to the plugin please create a new feature request describing the feature you would like to see.</p>
</pre>
</details>

</ul>
</details>

<p>Closing this issue as this is not a defect the plugin works as expected.</p>

<p>If you like a change to the plugin please create a new feature request describing the feature you would like to see.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a class="jira-relationship" href="https://issue-redirect.jenkins.io/issue/13710">JENKINS-13710</a>]</i>
